CalendarStorageException when infinite RRULE + RDATE present

### Discussed in https://github.com/bitfireAT/davx5-ose/discussions/402

<div type='discussions-op-text'>

<sup>Originally posted by **navid-zamani** September 10, 2023</sup>
Since there is no place to report bugs, I’m forced to post this here in a free format:

I get the above mentioned error when trying to sync.  
I suspect it was caused by a bug in Lightning, but I switched to KOrganzizer as a result and it lets me edit and save the event just fine, which I did, to make sure the event on the Radicale server was valid. Lightning is fine with it now too (but I set it to read-only there, to avoid Lightning messing up some more).

I’ve added the debug info ZIP below, and also added the ICS file from the server to it.
[davx5-debug.zip](https://github.com/bitfireAT/davx5-ose/files/12568152/davx5-debug.zip)
Everything starting with “My…”  is anonymized manually.

I am unable to find anything wrong with the ICS file, and since KOrganizer, Radicale, Thunderbird and an online validator handle it fine now, but DAVx⁵ (or more correctly, Android’s calendar provider or something) does not, all I can do it post it here and hope somebody more experienced can see what’s wrong.

It’s bad because I can’t use my calendar on my phone because of it right now. :/</div>

More adb details:

```
java.lang.IllegalArgumentException: Unable to compute LAST_DATE
	at com.android.providers.calendar.CalendarProvider2.handleUpdateEvents(CalendarProvider2.java:4107)
	at com.android.providers.calendar.CalendarProvider2.updateInTransactionInner(CalendarProvider2.java:4356)
	at com.android.providers.calendar.CalendarProvider2.updateInTransaction(CalendarProvider2.java:4208)
	at com.android.providers.calendar.SQLiteContentProvider.update(SQLiteContentProvider.java:168)
	at com.android.providers.calendar.CalendarProvider2.update(CalendarProvider2.java:2322)
	at android.content.ContentProvider.update(ContentProvider.java:2045)
	at android.content.ContentProviderOperation.applyInternal(ContentProviderOperation.java:391)
	at android.content.ContentProviderOperation.apply(ContentProviderOperation.java:352)
	at com.android.providers.calendar.SQLiteContentProvider.applyBatch(SQLiteContentProvider.java:236)
	at com.android.providers.calendar.CalendarProvider2.applyBatch(CalendarProvider2.java:2301)
	at android.content.ContentProvider.applyBatch(ContentProvider.java:2644)
	at android.content.ContentProvider$Transport.applyBatch(ContentProvider.java:514)
	at android.content.ContentProviderNative.onTransact(ContentProviderNative.java:212)
	at android.os.Binder.execTransactInternal(Binder.java:1339)
	at android.os.Binder.execTransact(Binder.java:1275)
Caused by: com.android.calendarcommon2.DateException: No range end provided for a recurrence that has no UNTIL or COUNT.
	at com.android.calendarcommon2.RecurrenceProcessor.expand(RecurrenceProcessor.java:766)
	at com.android.calendarcommon2.RecurrenceProcessor.expand(RecurrenceProcessor.java:661)
	at com.android.calendarcommon2.RecurrenceProcessor.getLastOccurence(RecurrenceProcessor.java:130)
	at com.android.calendarcommon2.RecurrenceProcessor.getLastOccurence(RecurrenceProcessor.java:61)
	at com.android.providers.calendar.CalendarProvider2.calculateLastDate(CalendarProvider2.java:3221)
	at com.android.providers.calendar.CalendarProvider2.handleUpdateEvents(CalendarProvider2.java:4105)
	... 14 more

```

--- a/lib/src/androidTest/kotlin/at/bitfire/ical4android/AndroidEventTest.kt
+++ b/lib/src/androidTest/kotlin/at/bitfire/ical4android/AndroidEventTest.kt
@@ -20,7 +20,6 @@ import at.bitfire.ical4android.util.DateUtils
 import at.bitfire.ical4android.util.MiscUtils.asSyncAdapter
 import at.bitfire.ical4android.util.MiscUtils.closeCompat
 import net.fortuna.ical4j.model.*
-import net.fortuna.ical4j.model.Recur.Frequency
 import net.fortuna.ical4j.model.component.VAlarm
 import net.fortuna.ical4j.model.parameter.*
 import net.fortuna.ical4j.model.property.*

--- a/lib/src/androidTest/kotlin/at/bitfire/ical4android/AndroidEventTest.kt
+++ b/lib/src/androidTest/kotlin/at/bitfire/ical4android/AndroidEventTest.kt
@@ -20,6 +20,7 @@ import at.bitfire.ical4android.util.DateUtils
 import at.bitfire.ical4android.util.MiscUtils.asSyncAdapter
 import at.bitfire.ical4android.util.MiscUtils.closeCompat
 import net.fortuna.ical4j.model.*
+import net.fortuna.ical4j.model.Recur.Frequency
 import net.fortuna.ical4j.model.component.VAlarm
 import net.fortuna.ical4j.model.parameter.*
 import net.fortuna.ical4j.model.property.*
@@ -262,6 +263,21 @@ class AndroidEventTest {
         assertNull(values.get(Events.EVENT_END_TIMEZONE))
 
         assertEquals("${tzShanghai.id};20200601T123000,20200701T183000,20200702T183000,20200801T123000,20200802T123000", values.getAsString(Events.RDATE))
+    }
+
+    @Test
+    fun testBuildEvent_NonAllDay_DtEnd_NoDuration_Recurring_InfiniteRruleAndRdate() {
+        val values = buildEvent(false) {
+            dtStart = DtStart("20200601T123000", tzShanghai)
+            dtEnd = DtEnd("20200601T123000", tzVienna)
+            rRules += RRule(
+                Recur("FREQ=DAILY;INTERVAL=2")
+            )
+            rDates += RDate(DateList("20200701T123000,20200702T123000", Value.DATE_TIME, tzVienna))
+        }
+
+        assertNotNull(values.get(Events.RRULE))
+        assertNull(values.get(Events.RDATE))
     }
 
     @Test

--- a/lib/src/androidTest/kotlin/at/bitfire/ical4android/AndroidEventTest.kt
+++ b/lib/src/androidTest/kotlin/at/bitfire/ical4android/AndroidEventTest.kt
@@ -276,8 +276,8 @@ class AndroidEventTest {
             rDates += RDate(DateList("20200701T123000,20200702T123000", Value.DATE_TIME, tzVienna))
         }
 
-        assertNotNull(values.get(Events.RRULE))
         assertNull(values.get(Events.RDATE))
+        assertEquals("FREQ=DAILY;INTERVAL=2", values.get(Events.RRULE))
     }
 
     @Test

--- a/lib/src/main/kotlin/at/bitfire/ical4android/AndroidEvent.kt
+++ b/lib/src/main/kotlin/at/bitfire/ical4android/AndroidEvent.kt
@@ -791,15 +791,25 @@ abstract class AndroidEvent(
                 builder.withValue(Events.RRULE, null)
 
             if (event.rDates.isNotEmpty()) {
-                for (rDate in event.rDates)
-                    AndroidTimeUtils.androidifyTimeZone(rDate)
+                // ignore RDATEs when there's also an infinite RRULE [https://issuetracker.google.com/issues/216374004]
+                val infiniteRrule = event.rRules.any { rRule ->
+                    rRule.recur.count == -1 &&  // no COUNT AND
+                    rRule.recur.until == null   // no UNTIL
+                }
 
-                // Calendar provider drops DTSTART instance when using RDATE [https://code.google.com/p/android/issues/detail?id=171292]
-                val listWithDtStart = DateList()
-                listWithDtStart.add(dtStart.date)
-                event.rDates.addFirst(RDate(listWithDtStart))
+                if (infiniteRrule)
+                    Ical4Android.log.warning("Android can't handle infinite RRULE + RDATE [https://issuetracker.google.com/issues/216374004]; ignoring RDATE(s)")
+                else {
+                    for (rDate in event.rDates)
+                        AndroidTimeUtils.androidifyTimeZone(rDate)
 
-                builder.withValue(Events.RDATE, AndroidTimeUtils.recurrenceSetsToAndroidString(event.rDates, allDay))
+                    // Calendar provider drops DTSTART instance when using RDATE [https://code.google.com/p/android/issues/detail?id=171292]
+                    val listWithDtStart = DateList()
+                    listWithDtStart.add(dtStart.date)
+                    event.rDates.addFirst(RDate(listWithDtStart))
+
+                    builder.withValue(Events.RDATE, AndroidTimeUtils.recurrenceSetsToAndroidString(event.rDates, allDay))
+                }
             } else
                 builder.withValue(Events.RDATE, null)
 

